--- Lib/distutils/sysconfig.py.orig	2007-04-05 21:32:57.000000000 -0700
+++ Lib/distutils/sysconfig.py	2007-04-05 21:32:59.000000000 -0700
@@ -365,7 +365,7 @@
     if sys.platform == 'darwin' and g.has_key('CONFIGURE_MACOSX_DEPLOYMENT_TARGET'):
         cfg_target = g['CONFIGURE_MACOSX_DEPLOYMENT_TARGET']
         cur_target = os.getenv('MACOSX_DEPLOYMENT_TARGET', '')
-        if cfg_target != cur_target:
+        if cfg_target != cur_target and not cur_target=="":
             my_msg = ('$MACOSX_DEPLOYMENT_TARGET mismatch: now "%s" but "%s" during configure'
                 % (cur_target, cfg_target))
             raise DistutilsPlatformError(my_msg)